File: rust/src/request_signer.rs
// MIT No Attribution
// Copyright 2026 AWS
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

//! Request signer for IAM Roles Anywhere authentication (AWS SigV4 with X.509).

use chrono::{DateTime, Utc};
use percent_encoding::{utf8_percent_encode, AsciiSet, NON_ALPHANUMERIC};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;
use yubikey::piv::{self, AlgorithmId, SlotId};
use yubikey::YubiKey;

use crate::certificate_reader::KeyType;
use crate::error::YubiraError;

const SERVICE: &str = "rolesanywhere";
const ALGORITHM_PREFIX: &str = "AWS4-X509";

/// Characters that must NOT be percent-encoded in URI path segments.
/// Per RFC 3986: unreserved = ALPHA / DIGIT / "-" / "." / "_" / "~"
const URI_ENCODE_SET: &AsciiSet = &NON_ALPHANUMERIC
    .remove(b'-')
    .remove(b'.')
    .remove(b'_')
    .remove(b'~');

/// Get the algorithm string for the Authorization header.
pub fn algorithm_string(key_type: KeyType) -> &'static str {
    match key_type {
        KeyType::Rsa => "RSA-SHA256",
        KeyType::EcP256 => "ECDSA-SHA256",
        KeyType::EcP384 => "ECDSA-SHA384",
    }
}

/// Full algorithm identifier (e.g., "AWS4-X509-RSA-SHA256").
pub fn full_algorithm(key_type: KeyType) -> String {
    format!("{ALGORITHM_PREFIX}-{}", algorithm_string(key_type))
}

/// Compute SHA-256 hash of payload, returned as lowercase hex.
pub fn hash_payload(payload: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(payload);
    hex::encode(hasher.finalize())
}

/// Create canonical request string per AWS SigV4 spec.
pub fn create_canonical_request(
    method: &str,
    uri: &str,
    query_string: &str,
    headers: &BTreeMap<String, String>,
    payload_hash: &str,
) -> String {
    let canonical_method = method.to_uppercase();
    let canonical_uri = uri_encode_path(uri);
    let canonical_query = create_canonical_query_string(query_string);
    let (canonical_headers, signed_headers) = create_canonical_headers(headers);

    format!(
        "{canonical_method}\n{canonical_uri}\n{canonical_query}\n{canonical_headers}\n{signed_headers}\n{payload_hash}"
    )
}

/// Create string to sign per IAM Roles Anywhere spec.
pub fn create_string_to_sign(
    key_type: KeyType,
    timestamp: &DateTime<Utc>,
    region: &str,
    canonical_request: &str,
) -> String {
    let algorithm = full_algorithm(key_type);
    let request_datetime = timestamp.format("%Y%m%dT%H%M%SZ").to_string();
    let date_stamp = timestamp.format("%Y%m%d").to_string();
    let credential_scope = format!("{date_stamp}/{region}/{SERVICE}/aws4_request");

    let mut hasher = Sha256::new();
    hasher.update(canonical_request.as_bytes());
    let hashed_canonical = hex::encode(hasher.finalize());

    format!("{algorithm}\n{request_datetime}\n{credential_scope}\n{hashed_canonical}")
}

/// Sign data using the YubiKey private key.
pub fn sign(
    yubikey: &mut YubiKey,
    slot: SlotId,
    key_type: KeyType,
    data: &[u8],
) -> Result<Vec<u8>, YubiraError> {
    let algorithm_id = match key_type {
        KeyType::Rsa => AlgorithmId::Rsa2048,
        KeyType::EcP256 => AlgorithmId::EccP256,
        KeyType::EcP384 => AlgorithmId::EccP384,
    };

    // Build the input for piv::sign_data.
    // For ECDSA: raw hash bytes.
    // For RSA: full PKCS#1 v1.5 padded block (key-size bytes).
    // The YubiKey performs raw RSA — we must provide the complete padded message.
    let sign_input: Vec<u8> = match key_type {
        KeyType::EcP384 => {
            use sha2::Sha384;
            let mut hasher = Sha384::new();
            hasher.update(data);
            hasher.finalize().to_vec()
        }
        KeyType::EcP256 => {
            let mut hasher = Sha256::new();
            hasher.update(data);
            hasher.finalize().to_vec()
        }
        KeyType::Rsa => {
            let mut hasher = Sha256::new();
            hasher.update(data);
            let hash = hasher.finalize();

            // Build DER-encoded DigestInfo for SHA-256 (RFC 3447 §9.2)
            let digest_info_prefix: &[u8] = &[
                0x30, 0x31, 0x30, 0x0d, 0x06, 0x09,
                0x60, 0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01,
                0x05, 0x00, 0x04, 0x20,
            ];

            // PKCS#1 v1.5 signature padding for RSA-2048 (256 bytes total):
            // 0x00 || 0x01 || PS (0xFF bytes) || 0x00 || DigestInfo
            let key_len: usize = 256; // RSA-2048
            let t_len = digest_info_prefix.len() + hash.len(); // 19 + 32 = 51
            let ps_len = key_len - 3 - t_len; // 256 - 3 - 51 = 202

            let mut padded = Vec::with_capacity(key_len);
            padded.push(0x00);
            padded.push(0x01);
            padded.extend(std::iter::repeat(0xFF).take(ps_len));
            padded.push(0x00);
            padded.extend_from_slice(digest_info_prefix);
            padded.extend_from_slice(&hash);
            padded
        }
    };

    let signature = piv::sign_data(yubikey, &sign_input, algorithm_id, slot)
        .map_err(|e| YubiraError::Signing(e.to_string()))?;

    Ok(signature.to_vec())
}

/// Build the Authorization header value.
pub fn build_authorization_header(
    key_type: KeyType,
    cert_serial: &str,
    timestamp: &DateTime<Utc>,
    region: &str,
    signed_headers: &str,
    signature: &[u8],
) -> String {
    let algorithm = full_algorithm(key_type);
    let date_stamp = timestamp.format("%Y%m%d").to_string();
    let credential_scope = format!("{date_stamp}/{region}/{SERVICE}/aws4_request");

    format!(
        "{algorithm} Credential={cert_serial}/{credential_scope}, SignedHeaders={signed_headers}, Signature={}",
        hex::encode(signature)
    )
}

/// Get the signed headers string from a headers map.
pub fn get_signed_headers(headers: &BTreeMap<String, String>) -> String {
    headers.keys().cloned().collect::<Vec<_>>().join(";")
}

// --- Internal helpers ---

fn uri_encode_path(uri: &str) -> String {
    if uri.is_empty() {
        return "/".to_string();
    }

    let uri = if uri.starts_with('/') { uri.to_string() } else { format!("/{uri}") };

    uri.split('/')
        .map(|segment| {
            if segment.is_empty() {
                String::new()
            } else {
                utf8_percent_encode(segment, URI_ENCODE_SET).to_string()
            }
        })
        .collect::<Vec<_>>()
        .join("/")
}

fn create_canonical_query_string(query_string: &str) -> String {
    if query_string.is_empty() {
        return String::new();
    }

    let mut params: Vec<(&str, &str)> = query_string
        .split('&')
        .map(|param| {
            if let Some((k, v)) = param.split_once('=') {
                (k, v)
            } else {
                (param, "")
            }
        })
        .collect();

    params.sort_by(|a, b| a.0.cmp(&b.0).then(a.1.cmp(&b.1)));

    params
        .iter()
        .map(|(k, v)| format!("{k}={v}"))
        .collect::<Vec<_>>()
        .join("&")
}

fn create_canonical_headers(
    headers: &BTreeMap<String, String>,
) -> (String, String) {
    // BTreeMap is already sorted by key
    let canonical_lines: Vec<String> = headers
        .iter()
        .map(|(name, value)| {
            let trimmed = value.split_whitespace().collect::<Vec<_>>().join(" ");
            format!("{name}:{trimmed}")
        })
        .collect();

    let canonical_headers = format!("{}\n", canonical_lines.join("\n"));
    let signed_headers = headers.keys().cloned().collect::<Vec<_>>().join(";");

    (canonical_headers, signed_headers)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_hash_payload_empty() {
        assert_eq!(
            hash_payload(b""),
            "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
        );
    }

    #[test]
    fn test_uri_encode_path_empty() {
        assert_eq!(uri_encode_path(""), "/");
    }

    #[test]
    fn test_uri_encode_path_root() {
        assert_eq!(uri_encode_path("/"), "/");
    }

    #[test]
    fn test_uri_encode_path_sessions() {
        assert_eq!(uri_encode_path("/sessions"), "/sessions");
    }

    #[test]
    fn test_canonical_query_string_sorted() {
        assert_eq!(
            create_canonical_query_string("z=1&a=2&m=3"),
            "a=2&m=3&z=1"
        );
    }

    #[test]
    fn test_canonical_query_string_empty() {
        assert_eq!(create_canonical_query_string(""), "");
    }

    #[test]
    fn test_canonical_headers_sorted_lowercase() {
        let mut headers = BTreeMap::new();
        headers.insert("host".to_string(), "example.com".to_string());
        headers.insert("content-type".to_string(), "application/json".to_string());

        let (canonical, signed) = create_canonical_headers(&headers);

        assert!(canonical.starts_with("content-type:application/json\n"));
        assert_eq!(signed, "content-type;host");
    }

    #[test]
    fn test_full_algorithm_rsa() {
        assert_eq!(full_algorithm(KeyType::Rsa), "AWS4-X509-RSA-SHA256");
    }

    #[test]
    fn test_full_algorithm_ec256() {
        assert_eq!(full_algorithm(KeyType::EcP256), "AWS4-X509-ECDSA-SHA256");
    }

    #[test]
    fn test_full_algorithm_ec384() {
        assert_eq!(full_algorithm(KeyType::EcP384), "AWS4-X509-ECDSA-SHA384");
    }

    #[test]
    fn test_algorithm_string_rsa() {
        assert_eq!(algorithm_string(KeyType::Rsa), "RSA-SHA256");
    }

    #[test]
    fn test_algorithm_string_ec256() {
        assert_eq!(algorithm_string(KeyType::EcP256), "ECDSA-SHA256");
    }

    #[test]
    fn test_algorithm_string_ec384() {
        assert_eq!(algorithm_string(KeyType::EcP384), "ECDSA-SHA384");
    }

    #[test]
    fn test_hash_payload_known_value() {
        // SHA-256 of "hello"
        assert_eq!(
            hash_payload(b"hello"),
            "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824"
        );
    }

    #[test]
    fn test_hash_payload_json_body() {
        let body = b"{\"durationSeconds\":3600}";
        let hash = hash_payload(body);
        assert_eq!(hash.len(), 64); // SHA-256 hex is 64 chars
        assert!(hash.chars().all(|c| c.is_ascii_hexdigit()));
    }

    #[test]
    fn test_uri_encode_path_with_special_chars() {
        assert_eq!(uri_encode_path("/path with spaces"), "/path%20with%20spaces");
    }

    #[test]
    fn test_uri_encode_path_preserves_unreserved() {
        assert_eq!(uri_encode_path("/a-b_c.d~e"), "/a-b_c.d~e");
    }

    #[test]
    fn test_uri_encode_path_no_leading_slash() {
        assert_eq!(uri_encode_path("sessions"), "/sessions");
    }

    #[test]
    fn test_canonical_query_string_single_param() {
        assert_eq!(create_canonical_query_string("key=value"), "key=value");
    }

    #[test]
    fn test_canonical_query_string_no_value() {
        assert_eq!(create_canonical_query_string("key"), "key=");
    }

    #[test]
    fn test_canonical_headers_whitespace_trimming() {
        let mut headers = BTreeMap::new();
        headers.insert("host".to_string(), "  example.com  ".to_string());
        let (canonical, _) = create_canonical_headers(&headers);
        assert!(canonical.contains("host:example.com"));
    }

    #[test]
    fn test_canonical_headers_multi_space_collapse() {
        let mut headers = BTreeMap::new();
        headers.insert("host".to_string(), "a   b   c".to_string());
        let (canonical, _) = create_canonical_headers(&headers);
        assert!(canonical.contains("host:a b c"));
    }

    #[test]
    fn test_get_signed_headers() {
        let mut headers = BTreeMap::new();
        headers.insert("content-type".to_string(), "application/json".to_string());
        headers.insert("host".to_string(), "example.com".to_string());
        headers.insert("x-amz-date".to_string(), "20260312T000000Z".to_string());
        assert_eq!(get_signed_headers(&headers), "content-type;host;x-amz-date");
    }

    #[test]
    fn test_create_canonical_request_structure() {
        let mut headers = BTreeMap::new();
        headers.insert("host".to_string(), "rolesanywhere.us-east-1.amazonaws.com".to_string());
        headers.insert("content-type".to_string(), "application/json".to_string());

        let payload_hash = hash_payload(b"{}");
        let cr = create_canonical_request("POST", "/sessions", "", &headers, &payload_hash);

        let lines: Vec<&str> = cr.split('\n').collect();
        assert_eq!(lines[0], "POST");            // method
        assert_eq!(lines[1], "/sessions");        // URI
        assert_eq!(lines[2], "");                 // empty query string
        assert!(lines[3].starts_with("content-type:")); // first canonical header
        assert!(lines[4].starts_with("host:"));          // second canonical header
        // line 5 is empty (trailing newline from canonical headers block)
        assert_eq!(lines[6], "content-type;host");       // signed headers
        assert_eq!(lines[7], payload_hash);               // payload hash
    }

    #[test]
    fn test_create_string_to_sign_structure() {
        let ts = DateTime::parse_from_rfc3339("2026-03-12T10:30:00+00:00")
            .unwrap()
            .with_timezone(&Utc);

        let sts = create_string_to_sign(KeyType::Rsa, &ts, "us-east-1", "canonical-request");

        let lines: Vec<&str> = sts.split('\n').collect();
        assert_eq!(lines.len(), 4);
        assert_eq!(lines[0], "AWS4-X509-RSA-SHA256");
        assert_eq!(lines[1], "20260312T103000Z");
        assert_eq!(lines[2], "20260312/us-east-1/rolesanywhere/aws4_request");
        // lines[3] is the SHA-256 of "canonical-request"
        assert_eq!(lines[3].len(), 64);
    }

    #[test]
    fn test_build_authorization_header_format() {
        let ts = DateTime::parse_from_rfc3339("2026-03-12T10:30:00+00:00")
            .unwrap()
            .with_timezone(&Utc);

        let sig = vec![0xDE, 0xAD, 0xBE, 0xEF];
        let auth = build_authorization_header(
            KeyType::Rsa,
            "123456789",
            &ts,
            "us-east-1",
            "content-type;host;x-amz-date;x-amz-x509",
            &sig,
        );

        assert!(auth.starts_with("AWS4-X509-RSA-SHA256 Credential=123456789/"));
        assert!(auth.contains("20260312/us-east-1/rolesanywhere/aws4_request"));
        assert!(auth.contains("SignedHeaders=content-type;host;x-amz-date;x-amz-x509"));
        assert!(auth.contains("Signature=deadbeef"));
    }

    #[test]
    fn test_build_authorization_header_ec256() {
        let ts = Utc::now();
        let auth = build_authorization_header(
            KeyType::EcP256, "999", &ts, "eu-west-1", "host", &[0x01],
        );
        assert!(auth.starts_with("AWS4-X509-ECDSA-SHA256 "));
    }
}

File: python/yubira/request_signer.py
"""
MIT No Attribution
Copyright 2026 AWS

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
"""
"""Request signer module for IAM Roles Anywhere authentication requests."""

from dataclasses import dataclass
from datetime import datetime
from typing import Literal, Optional
from urllib.parse import quote
import hashlib

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding
from yubikit.piv import PivSession, SLOT, KEY_TYPE


class SigningError(Exception):
    """Raised when signing operation fails."""
    pass


class InvalidRequestError(Exception):
    """Raised when request parameters are invalid."""
    pass


@dataclass
class SignedRequest:
    """A signed HTTP request ready to be sent."""
    method: str
    uri: str
    headers: dict[str, str]
    body: bytes
    signature: bytes


SignatureAlgorithm = Literal[
    "RSA-SHA256", 
    "ECDSA-SHA256", 
]


class RequestSigner:
    """
    Creates and signs IAM Roles Anywhere authentication requests.
    
    Implements AWS SigV4 signing with X.509 certificates, using the YubiKey
    to perform cryptographic signing operations without exposing private keys.
    """
    
    SERVICE = "rolesanywhere"
    ALGORITHM_PREFIX = "AWS4-X509"
    
    def __init__(self, session: PivSession, slot: SLOT = SLOT.AUTHENTICATION):
        """
        Initialize the request signer.
        
        Args:
            session: An active PIV session from a connected YubiKey.
            slot: PIV slot containing the signing key. Defaults to AUTHENTICATION (9a).
        """
        self._session = session
        self._slot = slot
        self._key_type: Optional[KEY_TYPE] = None
        self._cached_algorithm: Optional[SignatureAlgorithm] = None
    
    def get_signature_algorithm(self) -> SignatureAlgorithm:
        """
        Determine signature algorithm based on key type in slot.
        
        Returns:
            The signature algorithm string (e.g., "SHA256withRSA", "SHA256withECDSA").
            
        Raises:
            SigningError: If the key type cannot be determined or is unsupported.
        """
        if self._cached_algorithm is not None:
            return self._cached_algorithm
        
        try:
            cert = self._session.get_certificate(self._slot)
            if cert is None:
                raise SigningError(f"No certificate found in slot {self._slot}")

            self._key_type = self._get_supported_key_type(cert.public_key())
            if self._key_type == KEY_TYPE.RSA2048:
                self._cached_algorithm = "RSA-SHA256"
            else:
                # IAM Roles Anywhere requires SHA-256 for both supported EC curves.
                self._cached_algorithm = "ECDSA-SHA256"

            return self._cached_algorithm
        except SigningError:
            raise
        except Exception as e:
            raise SigningError(f"Failed to determine signature algorithm: {e}") from e

    @staticmethod
    def _get_supported_key_type(public_key) -> KEY_TYPE:
        """Return the stable cross-language PIV key type or fail closed."""
        try:
            key_type = KEY_TYPE.from_public_key(public_key)
        except ValueError as e:
            raise SigningError(str(e)) from e

        if key_type == KEY_TYPE.RSA2048:
            return key_type
        if key_type in (KEY_TYPE.RSA1024, KEY_TYPE.RSA3072, KEY_TYPE.RSA4096):
            raise SigningError(
                f"Unsupported RSA key size: {public_key.key_size}. "
                "Yubira currently supports RSA-2048; use P-384 for a "
                "higher-strength cross-language configuration."
            )
        if key_type in (KEY_TYPE.ECCP256, KEY_TYPE.ECCP384):
            return key_type
        raise SigningError(f"Unsupported key type: {type(public_key).__name__}")

    def create_canonical_request(
        self,
        method: str,
        uri: str,
        query_string: str,
        headers: dict[str, str],
        payload_hash: str
    ) -> str:
        """
        Create canonical request string per AWS SigV4 spec.
        
        The canonical request format is:
        HTTPRequestMethod + '\n' +
        CanonicalURI + '\n' +
        CanonicalQueryString + '\n' +
        CanonicalHeaders + '\n' +
        SignedHeaders + '\n' +
        HashedPayload
        
        Args:
            method: HTTP method (GET, POST, etc.) - will be uppercased.
            uri: URI path (e.g., "/sessions").
            query_string: Query string without leading '?' (can be empty).
            headers: Dictionary of header names to values.
            payload_hash: SHA-256 hash of the request payload (hex-encoded).
            
        Returns:
            The canonical request string.
        """
        # 1. HTTP method (uppercase)
        canonical_method = method.upper()
        
        # 2. Canonical URI (URI-encoded path)
        canonical_uri = self._uri_encode_path(uri)
        
        # 3. Canonical query string (sorted by parameter name)
        canonical_query = self._create_canonical_query_string(query_string)
        
        # 4. Canonical headers (sorted, lowercase names, trimmed values)
        canonical_headers, signed_headers = self._create_canonical_headers(headers)
        
        # 5. Combine all parts
        canonical_request = "\n".join([
            canonical_method,
            canonical_uri,
            canonical_query,
            canonical_headers,
            signed_headers,
            payload_hash
        ])
        
        return canonical_request
    
    def _uri_encode_path(self, uri: str) -> str:
        """
        URI-encode the path component.
        
        Args:
            uri: The URI path to encode.
            
        Returns:
            URI-encoded path with '/' preserved.
        """
        if not uri:
            return "/"
        
        # Normalize path - ensure it starts with /
        if not uri.startswith("/"):
            uri = "/" + uri
        
        # IAM Roles Anywhere requires each non-empty path segment to be
        # URI-encoded twice. The second pass encodes percent signs emitted by
        # the first pass while preserving RFC 3986 unreserved characters.
        segments = uri.split("/")
        encoded_segments = []
        for segment in segments:
            if segment:
                encoded_once = quote(segment, safe="-_.~")
                encoded_segments.append(quote(encoded_once, safe="-_.~"))
            else:
                encoded_segments.append("")
        
        return "/".join(encoded_segments) or "/"
    
    def _create_canonical_query_string(self, query_string: str) -> str:
        """
        Create a canonical query string from an already URL-encoded query.

        Parameters are sorted by encoded name and value. IAM Roles Anywhere
        additionally requires equals signs in parameter values to be encoded
        twice, so both raw ``=`` and existing ``%3D`` forms become ``%253D``.
        """
        if not query_string:
            return ""
        
        params = []
        for param in query_string.split("&"):
            if "=" in param:
                key, value = param.split("=", 1)
            else:
                key, value = param, ""

            canonical_value = (
                value.replace("%3D", "%253D")
                .replace("%3d", "%253D")
                .replace("=", "%253D")
            )
            params.append((key, canonical_value))

        params.sort(key=lambda item: (item[0], item[1]))
        return "&".join(f"{key}={value}" for key, value in params)
    
    def _create_canonical_headers(self, headers: dict[str, str]) -> tuple[str, str]:
        """
        Create canonical headers and signed headers string.
        
        Args:
            headers: Dictionary of header names to values.
            
        Returns:
            Tuple of (canonical_headers, signed_headers).
            canonical_headers: Sorted headers with lowercase names and trimmed values.
            signed_headers: Semicolon-separated list of lowercase header names.
        """
        # Convert to lowercase keys and trim values
        normalized = {}
        for name, value in headers.items():
            lower_name = name.lower()
            # Trim leading/trailing whitespace and collapse internal whitespace
            trimmed_value = " ".join(value.split())
            normalized[lower_name] = trimmed_value
        
        # Sort by header name
        sorted_names = sorted(normalized.keys())
        
        # Build canonical headers (each header ends with newline)
        canonical_lines = []
        for name in sorted_names:
            canonical_lines.append(f"{name}:{normalized[name]}")
        canonical_headers = "\n".join(canonical_lines) + "\n"
        
        # Build signed headers
        signed_headers = ";".join(sorted_names)
        
        return canonical_headers, signed_headers

    def create_string_to_sign(
        self,
        timestamp: datetime,
        region: str,
        canonical_request: str
    ) -> str:
        """
        Create string to sign per IAM Roles Anywhere spec.
        
        The string to sign format is:
        Algorithm + '\n' +
        RequestDateTime + '\n' +
        CredentialScope + '\n' +
        HashedCanonicalRequest
        
        Args:
            timestamp: Request timestamp (UTC).
            region: AWS region (e.g., "us-east-1").
            canonical_request: The canonical request string.
            
        Returns:
            The string to sign.
        """
        # Get algorithm
        algorithm = f"{self.ALGORITHM_PREFIX}-{self.get_signature_algorithm()}"
        
        # Format timestamp as ISO8601 basic format
        request_datetime = timestamp.strftime("%Y%m%dT%H%M%SZ")
        
        # Create credential scope
        date_stamp = timestamp.strftime("%Y%m%d")
        credential_scope = f"{date_stamp}/{region}/{self.SERVICE}/aws4_request"
        
        # Hash the canonical request
        hashed_canonical = hashlib.sha256(canonical_request.encode("utf-8")).hexdigest()
        
        # Combine all parts
        string_to_sign = "\n".join([
            algorithm,
            request_datetime,
            credential_scope,
            hashed_canonical
        ])
        
        return string_to_sign
    
    def sign(self, data: bytes) -> bytes:
        """
        Sign data using YubiKey private key.
        
        PIN must be verified before calling this method.
        
        Args:
            data: The data to sign.
            
        Returns:
            The signature bytes.
            
        Raises:
            SigningError: If signing fails.
        """
        try:
            self.get_signature_algorithm()
            key_type = self._key_type
            if key_type is None:
                raise SigningError("Signing key type was not initialized")

            # IAM Roles Anywhere requires SHA-256 for all supported key types.
            hash_alg = hashes.SHA256()

            if key_type == KEY_TYPE.RSA2048:
                signature = self._session.sign(
                    self._slot,
                    key_type,
                    data,
                    hash_alg,
                    padding.PKCS1v15()
                )
            elif key_type in (KEY_TYPE.ECCP256, KEY_TYPE.ECCP384):
                signature = self._session.sign(
                    self._slot,
                    key_type,
                    data,
                    hash_alg,
                )
            else:
                raise SigningError(f"Unsupported PIV key type: {key_type}")

            return signature
            
        except SigningError:
            raise
        except Exception as e:
            raise SigningError(f"Signing failed: {e}") from e
    
    def sign_request(
        self,
        method: str,
        uri: str,
        headers: dict[str, str],
        body: bytes,
        region: str,
        timestamp: datetime,
        query_string: str = ""
    ) -> SignedRequest:
        """
        Create and sign a complete request.
        
        This method:
        1. Computes the payload hash
        2. Creates the canonical request
        3. Creates the string to sign
        4. Signs the string using the YubiKey
        5. Returns a SignedRequest with all headers including Authorization
        
        Args:
            method: HTTP method.
            uri: Request URI path.
            headers: Request headers (should include Host, Content-Type, X-Amz-Date).
            body: Request body bytes.
            region: AWS region.
            timestamp: Request timestamp (UTC).
            query_string: Query string without leading '?' (default empty).
            
        Returns:
            SignedRequest with signature and updated headers.
            
        Raises:
            SigningError: If signing fails.
        """
        # Compute payload hash
        payload_hash = hashlib.sha256(body).hexdigest()
        
        # Copy headers (don't add x-amz-content-sha256 - not needed for Roles Anywhere)
        request_headers = dict(headers)
        
        # Create canonical request (with query string)
        canonical_request = self.create_canonical_request(
            method=method,
            uri=uri,
            query_string=query_string,
            headers=request_headers,
            payload_hash=payload_hash
        )
        
        # Create string to sign
        string_to_sign = self.create_string_to_sign(
            timestamp=timestamp,
            region=region,
            canonical_request=canonical_request
        )
        
        # Sign the string
        signature = self.sign(string_to_sign.encode("utf-8"))
        
        # Create Authorization header
        algorithm = f"{self.ALGORITHM_PREFIX}-{self.get_signature_algorithm()}"
        date_stamp = timestamp.strftime("%Y%m%d")
        credential_scope = f"{date_stamp}/{region}/{self.SERVICE}/aws4_request"
        
        # Get signed headers list
        _, signed_headers = self._create_canonical_headers(request_headers)
        
        # Get certificate serial for credential (decimal string, not hex)
        cert = self._session.get_certificate(self._slot)
        cert_serial = str(cert.serial_number)
        
        authorization = (
            f"{algorithm} "
            f"Credential={cert_serial}/{credential_scope}, "
            f"SignedHeaders={signed_headers}, "
            f"Signature={signature.hex()}"
        )
        
        request_headers["Authorization"] = authorization
        
        return SignedRequest(
            method=method,
            uri=uri,
            headers=request_headers,
            body=body,
            signature=signature
        )
    
    @staticmethod
    def hash_payload(payload: bytes) -> str:
        """
        Compute SHA-256 hash of payload.
        
        Args:
            payload: The payload bytes to hash.
            
        Returns:
            Hex-encoded SHA-256 hash.
        """
        return hashlib.sha256(payload).hexdigest()
